cdap-app-fabric/src/main/java/co/cask/cdap/security/impersonation/DefaultUGIProvider.java

```diff
@@ -199,51 +241,160 @@ private File localizeKeytab(Location keytabLocation) throws IOException {
     return localKeytabFile.toFile();
   }
   
-  private Map<String, String> getRuntimeImpersonationProperties(NamespacedEntityId programId) {
-      if (programId instanceof ProgramRunId) {
-          ProgramRunId runId = ((ProgramRunId) programId);
-          RunRecordMeta runRecord = null;
-          
-          long sleepDelayMs = TimeUnit.MILLISECONDS.toMillis(100);
-          long startTime = System.currentTimeMillis();
-          long timeoutMs = TimeUnit.SECONDS.toMillis(10);
-          while (System.currentTimeMillis() - startTime < timeoutMs) {
-            try {
-                runRecord = this.store.getRun(runId);
-                if (runRecord != null) {
-                  break;
-                }
-                Thread.sleep(sleepDelayMs);
-            } catch (Exception e) {
-                LOG.info("Exception while trying to fecth RunRecord : " + e.getMessage());
-                return null;
-            }
-          }
+  private Map<String, String> getRuntimeImpersonationProperties(NamespacedEntityId entityId) {     
+    if (!(entityId instanceof ProgramRunId) && (!(entityId instanceof TwillRunProgramId))) {
+      LOG.debug("Entity Id not of type ProgramRunId or TwillRunProgramId, skip fetching runtime args");
+      return Collections.emptyMap();
+    }
+      
+    ProgramRunId programRunId;
+    
+    if (entityId instanceof TwillRunProgramId) {
+      programRunId = getProgramRunIdFromTwillRunProgramId(entityId);
+      if (programRunId == null) {
+        return Collections.emptyMap();
+      }
+    } else {
+      programRunId = ((ProgramRunId) entityId);
```

Comment:
Can it throw cast exception?

Reply:
No it wont. At the top of function (in which this code block exists), I have checked if 'programRunId' is not an instance of ProgramRunId or TwillRunProgramId. If not, I am returning from function.